File: solutions/fsharp/yacht/1/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One = 1
    | Two = 2
    | Three = 3
    | Four = 4
    | Five = 5
    | Six = 6

let score category dice =
    let counts = List.countBy id dice
    let count d = counts |> List.tryFind (fun (x, _) -> x = d) |> Option.defaultValue (d, 0) |> snd |> (*) (int d)
    let ds = List.distinct dice
    let sumOfDice = dice |> List.sumBy int
    
    match category with
    | Yacht -> if List.length ds = 1 then 50 else 0
    | Ones -> count Die.One
    | Twos -> count Die.Two
    | Threes -> count Die.Three
    | Fours -> count Die.Four
    | Fives -> count Die.Five
    | Sixes -> count Die.Six
    | Choice -> dice |> List.sumBy int
    | FullHouse ->
        if List.length counts <> 2 then 0
        elif List.exists (fun (_, c) -> c = 1) counts then 0
        else List.sumBy (fun (d, c) -> int(d) * c) counts
    | FourOfAKind ->
        counts |> List.tryFind (fun (_, c) -> c >= 4) |> Option.map (fun (d, _) -> int(d) * 4) |> Option.defaultValue 0
    | LittleStraight ->
        if sumOfDice = 15 then 30 else 0
    | BigStraight ->
        if sumOfDice = 20 then 30 else 0

File: solutions/javascript/yacht/1/yacht.js
//
// This is only a SKELETON file for the 'Yacht' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const score = (array, category) => {
  switch (category) {
        case 'yacht': return array.every(d => d === array[0]) ? 50 : 0
        case 'ones': return count(array, 1)
        case 'twos': return count(array, 2) * 2
        case 'threes': return count(array, 3) * 3
        case 'fours': return count(array, 4) * 4
        case 'fives': return count(array, 5) * 5
        case 'sixes': return count(array, 6) * 6
        case 'full house': return multiple(array, 3) && multiple(array, 2) ? sum(array) : 0
        case 'four of a kind': return (multiple(array, 4) || multiple(array, 5) || 0) * 4
        case 'little straight': return isStraight(array, 1) ? 30 : 0
        case 'big straight': return isStraight(array, 2) ? 30 : 0
        case 'choice': return sum(array)
    }
};


const count = (dice, n) => dice.filter(d => d === n).length
const multiple = (dice, n) => dice.find(d => dice.filter(f => f === d).length === n)
const isStraight = (dice, n) => dice.sort().every((d, i) => d === i + n)
const sum = dice => dice.reduce((acc, cur) => acc + cur, 0)
